design/open/uncomitted-gc.md

```diff
@@ -0,0 +1,119 @@
+# Uncommitted Garbage Collector
+
+## Motivation
+
+Uncommitted data which is no longer referenced (due to branch deletion, reset branch etc.) is not being deleted by lakeFS.
+This may result excessive storage usage and possible compliance issues.
+To solve this problem two approaches were suggested:
+1. A batch operation performed as part of an external process (GC)
+2. An online solution inside lakeFS
+
+Several attempts for an online solution have been made, most of which are documented [here](Link-to-hard-delete-proposal).
+This document will describe the **offline** GC process for uncommitted objects.
+
+Garbage collection of uncommitted data will be performed using the same principles of the current GC process.
+The basis for this is a GC Client (i.e. _Spark_ job) consuming objects information from both lakeFS and the storage adapter, 
+and using this information to determine which objects on the bucket can be deleted.
+
+For simplicity, in this document we will reduce this problem to the repository and branch level.
+
+## Design
+
+### Required changes by lakeFS
+
+The following are necessary changes in lakeFS in order to implement this proposal successfully.
+
+#### Objects Path Conventions
+
+1. Repository objects will be stored under the prefix `<storage_namespace>/repos/<repo_uid>/`
+2. For each repository branch, objects will be stored under the repo prefix with the path `branches/<branch_id>/`
```

Comment:
We should address create/delete/create of the same branch by using unique id also in the branch level.